Question interface

Raul9595 opened the pull request:
# Thank You for your contribution to the *PyJudge*

- [x] Added a question interface to submit questions
- [x] Added a contest interface to create contests

I have not changed the existing **Question** Model and have created a new Model **UploadedQuestion** that will handle the Questions submitted through the web interface

theSage21 commented on server.py:
```diff
@@ -175,6 +188,61 @@ def statistics():
+@app.get("/display/<id>")
+@app.post("/display/<id>")
+@login_required
+def displayQuestion(id):
```
ah. you'll have to replace the contest question url instead of creating a new one.

theSage21 commented on server.py:
```diff
@@ -71,7 +82,9 @@ class Meta:
 db.connect()
-db.create_tables([User, Session, Submission, ContestProblems, Contest, Question])
+db.create_tables(
+    [User, Session, Submission, ContestProblems, Contest, Question, UploadedQuestion]
```
why two question classes? Let's replace the old Question with your new one.

Raul9595 replied:
Until the code was properly reviewed, I dint want to interfere with the already existing code

theSage21 commented on server.py:
```diff
@@ -50,6 +50,17 @@ class Meta:
 
 
+class UploadedQuestion(Model):
+    question_text = TextField()
```
is this the test case input?

Raul9595 replied:
Yes this is the Test Case input

Raul9595 replied:
I have changed it to a more meaningful field name **test_case_input**

Raul9595 commented:
- I have removed the existing table and replaced it with my table. I have kept the tried to keep the test data intact as much as possible. 
- I have given more meaningful field names. 
- I have used the existing contest question URL instead of creating a new one.

theSage21 commented on server.py:
```diff
@@ -175,22 +181,65 @@ def statistics():
+@app.post("/questionInput")
+@login_required
+def questionInput():
+    userid = Session.get(Session.token == bottle.request.get_cookie("s_id")).user
```
```suggestion
    userid = bottle.request.session.user
```

theSage21 replied:
the `session` is added when you use `login_required`

theSage21 commented on server.py:
```diff
@@ -175,22 +181,65 @@ def statistics():
+            test_case_output=uploaded_answer,
+            question_statement=uploaded_statement,
+            author=userid,
+            created_date_time=time,
```
while defining the model you can supply a default value and the value can be a function. so this does not have to be specified.

theSage21 commented on server.py:
```diff
@@ -43,8 +44,11 @@ class Meta:
+    test_case_output = TextField()
+    question_statement = CharField()
     author = ForeignKeyField(User)
+    created_date_time = DateTimeField()
```
```suggestion
    created_date_time = DateTimeField(default=datetime.datetime.now)
```

theSage21 commented on server.py:
```diff
@@ -175,22 +181,65 @@ def statistics():
+            author=userid,
+            created_date_time=time,
+        )
+    except:
```
an empty except is not good

```suggestion
    except Exception as e:
```

theSage21 commented on server.py:
```diff
@@ -175,22 +181,65 @@ def statistics():
+            created_date_time=time,
+        )
+    except:
+        bottle.abort(500, "Error in inserting submission to database.")
```
```suggestion
        bottle.abort(500, str(e))
```

theSage21 commented on server.py:
```diff
@@ -101,13 +107,13 @@ class Meta:
-q5 = Question.get_or_create(q_no=5, author=test[0])
-q6 = Question.get_or_create(q_no=6, author=test[0])
-
+q1 = Question.get_or_create(test_case_input="1", test_case_output="1", question_statement="1", author=test[0], created_date_time=datetime.datetime.now())
```
this dummy data is starting to clutter the code now. How about we move it to another file? Now that we have a way to add questions via a HTTP request `requests.post` should be able to do it.

Let's have the flow like this:

```
python server.py &  # start this in background
python insert_dummy.py
```

Raul9595 commented:
I have made the above changes. The test data is in a new file called **insert_dummy.py** that has to be run from the console after the initial server is up.

theSage21 commented on insert_dummy.py:
```diff
@@ -0,0 +1,126 @@
+import datetime
```
umm. can we not touch the database outside the server code? Let's do this with `requests`?

Raul9595 replied:
Ok sure. Sorry for that. I'll change it.

Raul9595 replied:
But then at this point of time, without a database call I can only call the questions post method. There is the ContestProblems Model which gets populated in the server.py program which takes in question as an input. So when I run the server program, it will fail.
Am I going about this right or is there another way to it ?

theSage21 replied:
you mean there is no "contest" for you to create a "contestproblem" against right? If that is the case then you're missing an api somewhere. If we insert the questions via requests it essentially simulates what the browser would have to do. Perhaps this can be the flow:

1. add quesions without attaching them to contests
2. create a contest by selecting from a list of questions

Raul9595 commented:
I have added the interface for adding Contests too. I am yet to work on the **insert_dummy.py** file. Didn't get much time today and wanted to finish the Contest interface also.

theSage21 commented:
take your time. I'm also running a little busy these days.

Raul9595 commented:
I have added updated the **insert_dummy.py** file. Adding screenshots of the Add Question and Add Contest Interface.

Main Page :
![image](https://user-images.githubusercontent.com/16488421/58493821-65d1cc80-8191-11e9-935e-160e32622988.png)

Question Upload Page :
![image](https://user-images.githubusercontent.com/16488421/58493959-ae898580-8191-11e9-9684-427cf7fe9395.png)

Question List Page :
![image](https://user-images.githubusercontent.com/16488421/58494019-ce20ae00-8191-11e9-8f12-dc9b49e423c8.png)

Contest Creation Page :
![image](https://user-images.githubusercontent.com/16488421/58494057-e09ae780-8191-11e9-9f3c-af81168d48dc.png)

Raul9595 commented:
I actually wanted to start development #128 . Should I start working on the original PyJaipur master branch copy or the one which has been accepted here ?
